### DISCUSSION TITLE
How to only allow one None between two fields and set one based on the other

### DISCUSSION THREAD

**Discussion by gpserger** [pydantic/pydantic#9608] [Question] [answered]
Hey! Currently learning pydantic for the first time and hoping to know the best way to do this:

I have a class
```python
Class Job(BaseModel):
  scheduled_time: datetime,
  cron_string: str | None,
```
I want some custom logic for validating/initializing an instance of this class:
if schedule_time is None, it should use [croniter](https://pypi.org/project/croniter/)(cron_string).get_next(datetime) to populate scheduled_time.
if both scheduled_time is none and cron_string is none it should fail.

I've tried using @field_validator but it's not really working the way I want. 
Any ideas?
Is this method a dumb way to do things?
Thanks!

**Comment by chyok** [comment, marked as answer]
Hi @gpserger ,
you can try this:
```python
from croniter import croniter

from datetime import datetime
from pydantic import BaseModel, field_validator


class Job(BaseModel):
    cron_string: str | None
    scheduled_time: datetime | None

    @field_validator('scheduled_time', mode="before")
    @classmethod
    def populate_scheduled_time(cls, v, info):
        cron_string = info.data.get('cron_string')
        if v is None and cron_string is not None:
            return croniter(cron_string).get_next(datetime)
        elif v is None and cron_string is None:
            raise ValueError('Either `scheduled_time` or `cron_string` must be provided')
        return v


a = Job(cron_string="*/5 * * * *", scheduled_time=None)
print(a)
```

Validation is performed in the order the fields are defined, so you will need to move the cron_string field above the scheduled_time field in the class definition.

**Reply by gpserger** [reply]
Thank you that works
